tower-http/src/follow_redirect/mod.rs

```diff
@@ -0,0 +1,325 @@
+//! Middleware for following redirections.
+
+pub mod policy;
+
+use self::policy::{ActionKind, Attempt, Policy, Standard};
+use futures_core::ready;
+use http::{
+    header::LOCATION, HeaderMap, HeaderValue, Method, Request, Response, StatusCode, Uri, Version,
+};
+use http_body::Body;
+use iri_string::{
+    spec::UriSpec,
+    types::{RiAbsoluteString, RiReferenceStr},
+};
+use pin_project::pin_project;
+use std::{
+    convert::TryFrom,
+    future::Future,
+    mem,
+    pin::Pin,
+    str,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// [`Layer`] for retrying requests with a [`Service`] to follow redirection responses.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct FollowRedirectLayer<P = Standard> {
+    policy: P,
+}
+
+impl FollowRedirectLayer {
+    /// Create a new [`FollowRedirectLayer`] with a [`Standard`] redirection policy.
+    pub fn standard() -> Self {
+        Self::default()
+    }
+}
+
+impl<P> FollowRedirectLayer<P> {
+    /// Create a new [`FollowRedirectLayer`] with the given redirection [`Policy`].
+    pub fn new(policy: P) -> Self {
+        FollowRedirectLayer { policy }
+    }
+}
+
+impl<S, P> Layer<S> for FollowRedirectLayer<P>
+where
+    S: Clone,
+    P: Clone,
+{
+    type Service = FollowRedirect<S, P>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        FollowRedirect::new(inner, self.policy.clone())
+    }
+}
+
+/// Middleware that retries requests with a [`Service`] to follow redirection responses.
+#[derive(Clone, Copy, Debug)]
+pub struct FollowRedirect<S, P = Standard> {
+    inner: S,
+    policy: P,
+}
+
+impl<S> FollowRedirect<S> {
+    /// Create a new [`FollowRedirect`] with a [`Standard`] redirection policy.
+    pub fn standard(inner: S) -> Self {
+        Self::new(inner, Standard::default())
+    }
+}
+
+impl<S, P> FollowRedirect<S, P>
+where
+    P: Clone,
+{
+    /// Create a new [`FollowRedirect`] with the given redirection [`Policy`].
+    pub fn new(inner: S, policy: P) -> Self {
+        FollowRedirect { inner, policy }
+    }
+
+    /// Returns a new [`Layer`] that wraps services with a `FollowRedirect` middleware.
+    ///
+    /// [`Layer`]: tower_layer::Layer
+    pub fn layer(policy: P) -> FollowRedirectLayer<P> {
+        FollowRedirectLayer::new(policy)
+    }
```

Comment:
We should have `get_ref`, `get_mut`, and `into_inner` accessors here. You can generate them with `define_inner_service_accessors!()`.